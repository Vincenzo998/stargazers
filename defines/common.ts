export const USER = "user";
export const REPO = "repo";
