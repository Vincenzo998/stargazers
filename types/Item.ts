export type Item = { id: string; avatar_url: string; login: string };
