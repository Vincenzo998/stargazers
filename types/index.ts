export * from "./ListItemProps";
export * from "./Item";
export * from "./Info";
