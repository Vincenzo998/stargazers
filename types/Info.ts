// defines
import { REPO, USER } from "../defines";

export type Info = { [USER]: string; [REPO]: string };
