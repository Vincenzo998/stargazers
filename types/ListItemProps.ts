import { Item } from "./Item";

export type ListItemProps = { item: Item; index: number };
