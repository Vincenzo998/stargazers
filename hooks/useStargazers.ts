import { useCallback, useEffect, useState } from "react";
import axios from "axios";

// hooks
import { useDebounce } from "./useDebounce";

// types
import { Info } from "../types";

// defines
import { REPO, USER } from "../defines";

export const useStargazers = (info: Info) => {
  const [stargazers, setStargazers] = useState([]);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string>("");

  const debouncedInfo = useDebounce<Info>(info, 500);

  const getStargazers = useCallback(async () => {
    if (debouncedInfo[USER] && debouncedInfo[REPO]) {
      setLoading(true);
      try {
        const res = await axios.get("https://api.github.com/repos/" + debouncedInfo[USER] + "/" + debouncedInfo[REPO] + "/stargazers");
        const { data, status } = res || {};
        if (status === 200) {
          setStargazers(data);
          setError("");
        }
      } catch (err) {
        setError("Data not available");
      } finally {
        setLoading(false);
      }
    } else {
      setStargazers([]);
      setError("");
    }
  }, [debouncedInfo]);

  useEffect(() => {
    getStargazers();
  }, [getStargazers]);

  return {
    stargazers,
    loading,
    error,
  };
};
