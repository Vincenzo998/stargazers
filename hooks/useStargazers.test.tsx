import { renderHook } from "@testing-library/react-hooks";

// screens
import { useStargazers } from "./useStargazers";

// types
import { Info } from "../types";

// defines
import { REPO, USER } from "../defines";

test("useStargazers is invoked without crashing - default information", () => {
  const info: Info = {
    [USER]: "",
    [REPO]: "",
  };
  const {
    result: { current },
  } = renderHook(() => useStargazers(info));

  expect(current).toMatchSnapshot();
});
