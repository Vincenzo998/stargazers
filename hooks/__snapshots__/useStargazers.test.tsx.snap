// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStargazers is invoked without crashing - default information 1`] = `
{
  "error": "",
  "loading": false,
  "stargazers": [],
}
`;
